File: Ahorn/entities/DreamMirrorNoCS.jl
module VivHelperDreamMirrorNoCS
using ..Ahorn, Maple

@mapdef Entity "VivHelper/CutscenelessDreamMirror" DreamMirrorNCS(
    x::Integer, y::Integer, FrameType::String="default", GlassType::String="default", Broken::Bool=false, Reflection::String="InversePlayer"
)

const placements = Ahorn.PlacementDict(
   "Dream Mirror (No Cutscene, Semi-Custom) (Viv's Helper)" => Ahorn.EntityPlacement(
      DreamMirrorNCS
   )
)

const glassType = Dict{String, String}(
    "default" => "default",
    "Madeline" => "redMirror",
    "Badeline" => "purpMirror"
)

Ahorn.editingOptions(entity::DreamMirrorNCS) = Dict{String, Any}(
    "FrameType" => String["default", "copper", "gray", "purple", "shadow", "steel_gray", "warped1", "warped2"],
    "Reflection" => String["InversePlayer", "SameAsPlayer", "MaddyOnly", "BaddyOnly", "None"],
    "GlassType" => glassType
)

function frameFind(entity::DreamMirrorNCS)
    temp = get(entity.data, "FrameType", "default")
    frame = "VivHelper/MaddyBaddyMirror/dream_frame_"
    if temp == "default"
        return frame
    else
        return frame * temp
    end
end

function glassFind(entity::DreamMirrorNCS)
    Baddy2 = get(entity.data, "GlassType", "default")
    broke = get(entity.data, "Broken", false)
    broken = broke ? "01" : "00"
    if Baddy2 == "default"
        if broke
            return "objects/mirror/glassbreak09"
        else
            return "objects/mirror/glassbg"
        end
    else
        return "VivHelper/MaddyBaddyMirror/" * Baddy2 * broken
    end
end

function Ahorn.selection(entity::DreamMirrorNCS)
    x, y = Ahorn.position(entity)
    f = frameFind(entity)
    g = glassFind(entity)
    return Ahorn.coverRectangles([
        Ahorn.getSpriteRectangle(f, x, y, jx=0.5, jy=1.0),
        Ahorn.getSpriteRectangle(g, x, y, jx=0.5, jy=1.0)
    ])
end

const shine = "VivHelper/MaddyBaddyMirror/shine00"

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::DreamMirrorNCS, room::Maple.Room)
    frame = frameFind(entity)
    glass = glassFind(entity)

    Ahorn.drawSprite(ctx, glass, 0, 0, jx=0.5, jy=1.0)
    Ahorn.drawSprite(ctx, shine, 0, 0, jx=0.5, jy=1.0, alpha=0.5)
    Ahorn.drawSprite(ctx, frame, 0, 0, jx=0.5, jy=1.0)

end

end

File: Loenn/entities/DreamMirrorNoCS.lua
local drawableSpriteStruct = require("structs.drawable_sprite")

local dreamMirror = {}

dreamMirror.name = "VivHelper/CutscenelessDreamMirror"
dreamMirror.placements = {
    name = "normal",
    data = {
        FrameType = "default",
        GlassType = "default",
        Broken = false,
        Reflection = "InversePlayer"
    }
}
dreamMirror.fieldInformation = {
    FrameType = {fieldType = "string", options = {"default", "copper", "gray", "purple", "shadow", "steel_gray", "warped1", "warped2"}, editable = false},
    Reflection = {fieldType = "string", options = {"InversePlayer", "SameAsPlayer", "MaddyOnly", "BaddyOnly", "None"}, editable = false},
    GlassType = {fieldType = "string", options = {{"default", "default"}, {"Madeline", "redMirror"}, {"Badeline", "purpMirror"}}, editable = false}
}

function dreamMirror.sprite(room, entity)
    local sprites = {}
    local ft = entity.FrameType or "default"
    local gt = entity.GlassType or "default"
    if gt == "default" then gt = (entity.Broken and "objects/mirror/glassbreak09" or "objects/mirror/glassbg")
    else gt = "VivHelper/MaddyBaddyMirror/" .. gt .. (entity.Broken and "01" or "00") end
    local frameSprite = drawableSpriteStruct.fromTexture("VivHelper/MaddyBaddyMirror/dream_frame_" .. ft, entity)
    frameSprite:setJustification(0.5, 1.0)
    frameSprite.depth = 9000

    local glassSprite = drawableSpriteStruct.fromTexture(gt, entity)
    glassSprite:setJustification(0.5, 1.0)
    glassSprite.depth = 9500

    local shineSprite = drawableSpriteStruct.fromTexture("VivHelper/MaddyBaddyMirror/shine00", entity)
    shineSprite:setJustification(0.5, 1.0)
    shineSprite:setAlpha(0.5)
    shineSprite.depth = 9499

    table.insert(sprites, glassSprite) -- Order important for pre-placement
    table.insert(sprites, shineSprite)
    table.insert(sprites, frameSprite)

    return sprites
end

return dreamMirror